### PR TITLE
[IMP] mail, mail_bot: move notif request and OdooBot name into mail

### PR DESCRIPTION
**Pull request by phenix-factory**
task-2282334

upgrade: https://github.com/odoo/upgrade/pull/1700

### CODE REVIEW

File: addons/mail/models/mail_channel.py
Comment (alexkuhn) [inline addons/mail/models/mail_channel.py:997]: this should stay in `mail_bot` module

File: addons/mail/models/mail_channel.py
Comment (alexkuhn) [inline addons/mail/models/mail_channel.py:1037]: should stay in `mail_bot` module

File: addons/mail/models/mail_thread.py
Comment (alexkuhn) [inline addons/mail/models/mail_thread.py:1919]: `mail_bot` module

File: addons/mail/models/res_users.py
Comment (alexkuhn) [inline addons/mail/models/res_users.py:46]: `mail_bot` module

File: addons/mail/models/res_users.py
Comment (alexkuhn) [inline addons/mail/models/res_users.py:84]: `mail_bot` module

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (alexkuhn) [inline addons/mail/static/src/components/notification_list/notification_list.js:8]: alphabetical order

File: addons/mail/static/src/models/messaging/messaging.js
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging/messaging.js:127]: revert

File: addons/mail_bot/static/src/models/messaging_initializer/messaging_initializer.js
Comment (alexkuhn) [inline addons/mail_bot/static/src/models/messaging_initializer/messaging_initializer.js:35]: remove

File: addons/mail/views/res_users_views.xml
Comment (alexkuhn) [inline addons/mail/views/res_users_views.xml:59]: `mail_bot` module

File: addons/mail/views/assets.xml
Comment (alexkuhn) [inline addons/mail/views/assets.xml:164]: alphabetical order (by each path bit)

File: addons/mail/views/assets.xml
Comment (alexkuhn) [inline addons/mail/views/assets.xml:219]: alphabetical order (by each path bit)

File: addons/mail_bot/models/mail_channel.py
Comment (alexkuhn) [inline addons/mail_bot/models/mail_channel.py:23]: revert

File: addons/mail_bot/models/mail_thread.py
Comment (alexkuhn) [inline addons/mail_bot/models/mail_thread.py:12]: revert

File: addons/mail_bot/models/res_users.py
Comment (alexkuhn) [inline addons/mail_bot/models/res_users.py:29]: revert

**Comment by tde-banana-odoo** [comment]
What ? Why would you put odoo chatbot inside mail ?

**Comment by alexkuhn** [comment]
> What ? Why would you put odoo chatbot inside mail ?

Push notification and "System => Odoobot" are moved to `/mail`, the rest (chatbot) should stay in `/mail_bot`.

**Comment by tde-banana-odoo** [comment]
I see the whole "logic" of mailbot has been moved. Is it a mistake ?

**Comment by alexkuhn** [comment]
> I see the whole "logic" of mailbot has been moved. Is it a mistake ?

PR's still in review. I've just asked to revert all chatbot-related moves.

**Comment by tde-banana-odoo** [comment]
Ok, thanks for clarification :) . PR title made me have an heart attack. Twice a day is sufficient.

File: addons/mail/views/assets.xml
Comment (phenix-factory) [reply addons/mail/views/assets.xml:164]: Look like it's already the case...

File: addons/mail/views/assets.xml
Comment (phenix-factory) [reply addons/mail/views/assets.xml:219]: Look like it's already the case...

**Comment by phenix-factory** [comment]
PR updated

File: addons/mail/views/assets.xml
Comment (alexkuhn) [reply addons/mail/views/assets.xml:164]: not with other paths already in bundle

File: addons/mail/views/assets.xml
Comment (alexkuhn) [reply addons/mail/views/assets.xml:219]: not with other paths already in bundle

File: addons/mail/models/mail_channel.py
Comment (alexkuhn) [inline addons/mail/models/mail_channel.py:1035]: should stay in `mail_bot` (chatbot)

File: addons/mail/models/mail_thread.py
Comment (alexkuhn) [reply addons/mail/models/mail_thread.py:1920]: `_apply_logic` is used to generate answer from message post by current user, so chatbot-related code, so should be stay in `mail_bot`

File: addons/mail/views/assets.xml
Comment (alexkuhn) [reply addons/mail/views/assets.xml:164]: For instance, for components:
- ...
- **<script type="text/javascript" src="/mail/static/src/components/notification_alert/notification_alert.js"></script>**
- <script type="text/javascript" src="/mail/static/src/components/notification_group/notification_group.js"></script>
- <script type="text/javascript" src="/mail/static/src/components/notification_list/notification_list.js"></script>
- <script type="text/javascript" src="/mail/static/src/components/notification_popover/notification_popover.js">- </script>
- **<script type="text/javascript" src="/mail/static/src/components/notification_request/notification_request.js"></script>**
- ...

File: addons/mail/models/mail_thread.py
Comment (phenix-factory) [reply addons/mail/models/mail_thread.py]: Done

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (phenix-factory) [reply addons/mail/static/src/components/notification_list/notification_list.js]: Done

File: addons/mail/models/mail_channel.py
Comment (phenix-factory) [reply addons/mail/models/mail_channel.py:1036]: Done

**Comment by phenix-factory** [comment]
PR updated

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (alexkuhn) [inline addons/mail/static/src/components/notification_list/notification_list.js:147]: misindent

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (alexkuhn) [inline addons/mail/static/src/components/notification_list/notification_list.js:153]: here could push "odoobot has a request" notification before adding thread notifications. `unshift()` made sense with override because notifications required super call.

File: addons/mail/static/src/models/messaging_menu/messaging_menu.js
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_menu/messaging_menu.js:82]: `counter++;`

**Comment by alexkuhn** [comment]
migration-remove-mailbot-view
@seb-odoo to solve that, we should make changes of this PR in master before 13.4?

**Comment by seb-odoo** [comment]
This one is a really removed view, there needs to be a migration script.

**Comment by seb-odoo** [comment]
Oops no nevemind, `view_users_form_inherit_mail_notification_alert` is just an override that was moved into mail?
Migration script should be needed only to remove it from mail_bot, not to add it into mail (that part specifically has to be done in master if the bug is still there)

**Comment by seb-odoo** [comment]
Also `view_users_form_mail` is already inherited in mail, maybe it can be combined as to not have a new template in the first place?

**Comment by phenix-factory** [comment]
@alexkuhn I have created a PR for master https://github.com/odoo/odoo/pull/55616

**Comment by phenix-factory** [comment]
> Also `view_users_form_mail` is already inherited in mail, maybe it can be combined as to not have a new template in the first place?

The field Odoobot_state is part of mail_bot. It's in @alexkuhn review.

File: addons/mail/views/res_users_views.xml
Comment (alexkuhn) [inline addons/mail/views/res_users_views.xml:43]: > > Also `view_users_form_mail` is already inherited in mail, maybe it can be combined as to not have a new template in the first place?
> 
> The field Odoobot_state is part of mail_bot. It's in @alexkuhn review.

I think @seb-odoo meant `base.view_users_form_simple_modif` is already inherited in mail by `view_users_form_simple_modif_mail`, so could just add `xpath` on this latter inherit?

Comment (phenix-factory) [reply addons/mail/views/res_users_views.xml]: OK ! got it done :)

File: addons/mail/views/assets.xml
Comment (seb-odoo) [inline addons/mail/views/assets.xml:217]: Odoobot command style can stay in mail_bot

File: addons/mail_bot/views/res_users_views.xml
Comment (seb-odoo) [inline addons/mail_bot/views/res_users_views.xml:15]: revert this diff, this part stays here, no need to move it below on the file.

Just `view_users_form_inherit_mail_notification_alert` can be removed

**Review by seb-odoo — CHANGES_REQUESTED** [review]
quick check, not tested yet

File: addons/mail/models/res_partner.py
Comment (phenix-factory) [reply addons/mail/models/res_partner.py]: Your missing the `.env` for the ref. So `self.env.ref("base.partner_root')`.

File: addons/mail_bot/views/res_users_views.xml
Comment (seb-odoo) [inline addons/mail_bot/views/res_users_views.xml:41]: I don't understand why keep this diff. `view_users_form_inherit_mail_notification_alert` should be moved to `mail` and simply removed from here.

There is already `res_users_view_form` in `mail_bot` for the `odoobot_state` stuff.

Unless there is something I'm missing, there should be no "green" (+) diff on this file.

File: addons/test_mail/tests/test_mail_channel.py
Comment (seb-odoo) [inline addons/test_mail/tests/test_mail_channel.py:411]: This should not be part of `test_mail_channel` file, as it's just testing a method on partner.

Considering the currently existing test files, I think a new one for this would be appropriate.

eg. `test_partner_mention`


Also these tests need adaptation, as it is it not not clear what is their goal, and where the result is coming from.

I think I retract my idea of adding these test on this PR. You can just remove them here and keep the diff for later.

We should make a new task for that purpose to properly and fully testing mentions python side instead.

File: addons/mail/models/res_partner.py
Comment (phenix-factory) [reply addons/mail/models/res_partner.py:115]: My IDE killed a trailling white space (good boy) !

**Comment by phenix-factory** [comment]
PR updated :)

File: addons/mail/models/res_partner.py
Comment (seb-odoo) [inline addons/mail/models/res_partner.py:121]: should use `odoobot.read(fields)[0]` to be more consistent with the rest of the method.

File: addons/mail/static/src/models/messaging/messaging.js
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging/messaging.js:117]: alphabetical

File: addons/mail_bot/views/assets.xml
Comment (seb-odoo) [inline addons/mail_bot/views/assets.xml:25]: this should not be removed

File: addons/mail_bot/views/assets.xml
Comment (seb-odoo) [inline addons/mail_bot/views/assets.xml:32]: the corresponding file was not deleted

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (seb-odoo) [inline addons/mail/static/src/components/notification_list/notification_list.js:147]: minimal diff: `unshift` into `notifications` instead of adding `res`

**Review by seb-odoo — CHANGES_REQUESTED** [review]
I'll apply them myself but here are the last changes I noticed just so you know

File: addons/mail/static/src/components/messaging_menu/messaging_menu_tests.js
Comment (seb-odoo) [inline addons/mail/static/src/components/messaging_menu/messaging_menu_tests.js:833]: also these mocks were removed, unnoticed rebase effect

File: addons/mail/static/src/models/messaging/messaging.js
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging/messaging.js:31]: Could be turned into a computed field... just unfortunate need to manually re-compute because `this.env.browser` doesn't sync with model fields... Could be left like this for the moment.

File: addons/mail_bot/__manifest__.py
Comment (alexkuhn) [inline addons/mail_bot/__manifest__.py:6]: do we really need to bump version?

File: addons/mail/static/src/components/notification_list/notification_list.xml
Comment (alexkuhn) [inline addons/mail/static/src/components/notification_list/notification_list.xml:39]: don't like relying on magic variables of `t-foreach`, but we know we have to re-work notifications so could be done later.

**Review by alexkuhn — APPROVED** [review]
LGTM

File: addons/mail_bot/__manifest__.py
Comment (seb-odoo) [reply addons/mail_bot/__manifest__.py:6]: Have to do a migration for removed inherit, this is guideline for when we do a migration on a module.

File: addons/mail/static/src/models/messaging/messaging.js
Comment (seb-odoo) [reply addons/mail/static/src/models/messaging/messaging.js:31]: Yep, needs to be recomputed every time specifically. Can't be a compute if we can't express dependencies.

File: addons/mail/static/src/components/notification_list/notification_list.xml
Comment (seb-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.xml:39]: I just removed the `name` from that tag btw, this is exactly like before.

This is not related to notifications though, it's just a separator here.

The only "way" to get rid of this foreach variable is to have the separator modeled as part of the list.

**Comment by tde-banana-odoo** [comment]
Why 13.4 ? Why not master ?

**Comment by KangOl** [comment]
Should **not** target `saas-13.4`. Please retarget this PR to `master` (or `saas-13.5`, it's still tolerated)

**Comment by seb-odoo** [comment]
Ok, probably can't retarget with new runbot, will make a new 13.5 PR.

**Comment by tde-banana-odoo** [comment]
Actually a saas-13.5-something branch will not merge any migration script. You have to wait for master forward-port and attach an upgrade branch to it, or merge in master unless you think your PR is absolutely critical for v14 (13.5).

**Comment by seb-odoo** [comment]
Yes the PR was asked for v14 (it was for 13.4 originally but we were late on it).

I think we can merge without trouble in 13.5 because runbot has proper CI even without the upgrade script. At least it was fine in 13.4, hopefully 13.5 will be the same.
Upgrade branch itself can target master as usual.

**Comment by tde-banana-odoo** [comment]
What you can do is make a saas-13.5-odoobot branch, merge it, wait for forward port master-saas-13.5-odoobot-oin-fw, and create a master branch with same name than the forward port before merging it. That way master forward port lands with its migration automatically.

**Comment by seb-odoo** [comment]
new PR: https://github.com/odoo/odoo/pull/56584

**Comment by seb-odoo** [comment]
> What you can do is make a saas-13.5-odoobot branch, merge it, wait for forward port master-saas-13.5-odoobot-oin-fw, and create a master branch with same name than the forward port before merging it. That way master forward port lands with its migration automatically.

This looks like a really unfortunate runbot/mergebot workaround. I don't need the upgrade script to be merged with the forward-port but with the original PR...

Anyway I created/updated the other PRs hopefully this will work.

**Comment by tde-banana-odoo** [comment]
Makes no sense to merge a saas-13.5 branch on a not-versioned repository whose only branch is master. The other solution would be to have a versioned upgrade repository but this is not currently planned as far as I know.

**Pull request closed without merging**
